### PR TITLE
Fix variable suggestions for KafkaSourceFactory with explicit key-val…

### PR DESCRIPTION
**Pull request by gskrobisz**
…ue types.

### CODE REVIEW

File: engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaSourceFactory.scala [hunk 1]
```diff
@@ -42,10 +42,26 @@ class KafkaSourceFactory[K: ClassTag, V: ClassTag](deserializationSchemaFactory:
...
                                                    processObjectDependencies: ProcessObjectDependencies)
-  extends FlinkSourceFactory[ConsumerRecord[K, V]] with SingleInputGenericNodeTransformation[FlinkSource[ConsumerRecord[K, V]]] with WithCachedTopicsExistenceValidator {
+  extends FlinkSourceFactory[ConsumerRecord[K, V]]
+    with SingleInputGenericNodeTransformation[FlinkSource[ConsumerRecord[K, V]]]
+    with WithCachedTopicsExistenceValidator
+    with WithExplicitPossibleProducedVariableTypes {
 
   protected val topicNameSeparator = ","
 
+  protected val keyTypingResult: TypedClass = Typed.typedClass[K]
```
Comment (lciolecki) [inline engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaSourceFactory.scala:52]: Why you assume there always be typed class?

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala [hunk 2]
```diff
@@ -32,7 +32,10 @@ class DefinitionExtractor[T](methodDefinitionExtractor: MethodDefinitionExtracto
     (obj match {
       //TODO: how validators/editors in NodeConfig should be handled for GenericNodeTransformation/WithExplicitMethodToInvoke?
       case e:GenericNodeTransformation[_] =>
-        val returnType = if (e.nodeDependencies.contains(OutputVariableNameDependency)) Unknown else Typed[Void]
+        val returnType = e match {
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala:35]: we shouldn't do it here. This should be possible return type of this component - current usage with passing union of key and value types does not make sense here. Let's rather use it in TypesInformation.extract method

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/WithExplicitPossibleProducedVariableTypes.scala [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package pl.touk.nussknacker.engine.definition
+
+import pl.touk.nussknacker.engine.api.typed.typing.TypedClass
+
+trait WithExplicitPossibleProducedVariableTypes {
+  def possibleVariableClasses: Set[TypedClass]
```
Comment (mproch) [inline engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/WithExplicitPossibleProducedVariableTypes.scala:6]: let's rename it somehow. These are not necessarily exact variable classes. Maybe sth along the lines of: 
`trait DeclaringExplicitTypesToExtract {
   def typesToExtract: Set[TypedClass]
}
`?

**Review by mproch — CHANGES_REQUESTED** [review]

File: engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaSourceFactory.scala (hunk 1 above)
Comment (gskrobisz) [reply engine/flink/kafka-util/src/main/scala/pl/touk/nussknacker/engine/kafka/source/KafkaSourceFactory.scala:52]: Here K and V are usually a specific class or some generic Map/TypedMap. For specific classes this is the same as previous assumption with Typed[K] and Typed[V]. For generic Map/TypedMap the returning type is overwritten by TypedObjectTypingResult which has explicit fields definition.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala (hunk 2 above)
Comment (arkadius) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala:35]: By component you mean "Component" class? IMO we shouldn't do it too much generic. Take a look that this `possibleVariableClasses` takes no arguments. If we would like to do it correctly, we should probable traverse the whole scenario (not definition) and take a look what classes are returned. Also "codefull" approach of declaring components is deprecated. Isn't it?

Comment (gskrobisz) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala]: I moved it to `TypesInformation.extract`.

File: engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/WithExplicitPossibleProducedVariableTypes.scala (hunk 3 above)
Comment (gskrobisz) [reply engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/WithExplicitPossibleProducedVariableTypes.scala]: Renamed.

**Review by arkadius — APPROVED** [review]
please move the trait to api module

**Review by mproch — APPROVED** [review]

**Pull request merged**